tests/csapi/upload_keys_test.go

```diff
@@ -0,0 +1,208 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/matrix-org/complement/runtime"
+	"github.com/tidwall/gjson"
+	"maunium.net/go/mautrix/crypto/olm"
+)
+
+func TestUploadKey(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	deviceKeys, oneTimeKeys := generateKeys(t, alice, 1)
+	t.Run("Parallel", func(t *testing.T) {
+		// sytest: Can upload device keys
+		t.Run("Can upload device keys", func(t *testing.T) {
+			reqBody := client.WithJSONBody(t, map[string]interface{}{
+				"device_keys":   deviceKeys,
+				"one_time_keys": oneTimeKeys,
+			})
+			resp := alice.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "keys", "upload"}, reqBody)
+			must.MatchResponse(t, resp, match.HTTPResponse{
+				StatusCode: http.StatusOK,
+				JSON: []match.JSON{
+					match.JSONKeyEqual("one_time_key_counts.signed_curve25519", float64(1)),
+				},
+			})
+		})
+
+		// sytest: Rejects invalid device keys
+		t.Run("Rejects invalid device keys", func(t *testing.T) {
```

Comment:
Please please please keep useful sytest comments when porting things over. You don't explain in this test how this request is invalid :/ 

The sytest is clearer: `# algorithms, keys and signatures are required fields, but missing`